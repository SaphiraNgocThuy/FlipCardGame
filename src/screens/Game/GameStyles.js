import styled from 'styled-components/native';
import { getStatusBarHeight } from '../../utils/helper';
import { padding } from '../../utils/commonStyles';

export const GameView = styled.View`
  padding-top: ${getStatusBarHeight()}px;
  flex: 1;
`;

export const RowView = styled.View`
  padding-top: ${padding.sm}px;
  flex-direction: row;
  justify-content: space-between;
`;

export const CardsView = styled.View`
  padding: ${padding.sm}px;
  flex-direction: row;
  flex-wrap: wrap;
  flex-grow: 1;
  align-content: stretch;
`;
