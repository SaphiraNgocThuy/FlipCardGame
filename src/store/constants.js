export const FLIP_ALL = "FLIP_ALL";
export const GET_NUMBERS = "GET_NUMBERS";
export const PRESS_CARD = "PRESS_CARD";
export const FLIP_OPENED = "FLIP_OPENED";
