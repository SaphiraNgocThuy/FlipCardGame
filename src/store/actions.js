import { FLIP_ALL, PRESS_CARD, FLIP_OPENED, GET_NUMBERS } from "./constants";

export const restart = () => async (dispatch) => {
  dispatch({ type: FLIP_ALL });
  setTimeout(() => {
    dispatch({ type: GET_NUMBERS });
  }, 30);
};

export const pressCard = (title, id) => async (dispatch) =>
  dispatch({ type: PRESS_CARD, title, id });

export const flipOpened = () => async (dispatch) =>
  dispatch({ type: FLIP_OPENED });
