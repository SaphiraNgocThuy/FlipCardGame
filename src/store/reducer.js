import { FLIP_ALL, PRESS_CARD, FLIP_OPENED, GET_NUMBERS } from './constants';
import { getNumbers, getBackground } from '../utils/helper';
import images from '../assets';

export const initialState = {
  steps: 0,
  numbers: getNumbers(),
  firstSelected: null,
  secondSelected: null,
  resolved: [],
  background: images.fruit.strawberry,
};

const reducer = (state = initialState, action) => {
  const newState = { ...state };
  const { firstSelected, secondSelected, resolved, numbers } = newState;

  switch (action.type) {
    case PRESS_CARD:
      if (!firstSelected) {
        newState.firstSelected = { ...action };
      } else if (!secondSelected) {
        newState.secondSelected = { ...action };
      } else {
        if (firstSelected.title === secondSelected.title) {
          newState.resolved = [
            ...resolved,
            firstSelected.id,
            secondSelected.id,
          ];
        }
        newState.firstSelected = { ...action };
        newState.secondSelected = null;
      }
      newState.steps += 1;
      return newState;

    case FLIP_OPENED:
      if (firstSelected && secondSelected) {
        if (firstSelected.title === secondSelected.title) {
          newState.resolved = [
            ...resolved,
            firstSelected.id,
            secondSelected.id,
          ];
        }
        newState.firstSelected = null;
        newState.secondSelected = null;
      }
      return newState;

    case FLIP_ALL:
      return {
        ...initialState,
        numbers,
        background: images.fruit[getBackground()],
      };
    case GET_NUMBERS:
      return {
        ...state,
        numbers: getNumbers(),
      };

    default:
      return state;
  }
};
export default reducer;
