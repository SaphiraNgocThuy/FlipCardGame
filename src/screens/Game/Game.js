import React, { memo, useMemo, useCallback } from "react";
import { useSelector, useDispatch } from "react-redux";
import { restart, pressCard, flipOpened } from "../../store/actions";
import { GameView, RowView, CardsView } from "./GameStyles";
import CurvedButton from "../../components/Button/CurvedButton";
import CurvedBadge from "../../components/Badge/Badge";
import Card from "../../components/Card/Card";
import CongratsModal from "../../components/Modal/CongratsModal";

const Game = () => {
  const dispatch = useDispatch();
  const {
    steps,
    numbers,
    resolved,
    firstSelected,
    secondSelected,
  } = useSelector((state) => state);
  const win = useMemo(() => resolved.length === 12, [resolved]);

  const onCardPress = useCallback((title, id) => {
    dispatch(pressCard(title, id));
  }, []);

  const onFlipOpened = useCallback(() => {
    setTimeout(() => {
      if (secondSelected) dispatch(flipOpened());
    }, 1000);
  }, [secondSelected]);

  const renderCards = useMemo(
    () => [
      numbers.map((title, id) => {
        const isOpened =
          (firstSelected && firstSelected.id === id) ||
          (secondSelected && secondSelected.id === id) ||
          resolved.includes(id);

        return (
          <Card
            key={id}
            {...{ title, id, isOpened, onCardPress, onFlipOpened }}
          />
        );
      }),
    ],
    [numbers, resolved, firstSelected, secondSelected]
  );

  const onRestart = useCallback(() => dispatch(restart()), []);

  return (
    <GameView>
      <RowView>
        <CurvedButton title="Restart" onPress={onRestart} left />
        <CurvedBadge {...{ steps }} />
      </RowView>
      <CardsView>{renderCards}</CardsView>
      <CongratsModal {...{ steps }} visible={win} onPress={onRestart} />
    </GameView>
  );
};

export default memo(Game);
