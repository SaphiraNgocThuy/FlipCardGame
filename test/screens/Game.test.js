import React from 'react';
import renderer from 'react-test-renderer';
import * as redux from 'react-redux';
import { render, fireEvent } from 'react-native-testing-library';
import Game from '../../src/screens/Game/Game';
import { initialState } from '../../src/store/reducer';

it('renders correctly', () => {
  const useDispatch = jest.spyOn(redux, 'useDispatch');
  useDispatch.mockReturnValue(jest.fn());

  const useSelector = jest.spyOn(redux, 'useSelector');
  useSelector.mockReturnValue({
    ...initialState,
    numbers: [1, 1, 2, 2, 3, 4, 5, 5, 6, 6],
  });

  const tree = renderer.create(<Game />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should handle onRestart', () => {
  const dispatch = jest.fn();
  const useDispatch = jest.spyOn(redux, 'useDispatch');
  useDispatch.mockReturnValue(dispatch);
  const rendered = render(<Game />);
  const restartButton = rendered.getByText('Restart');
  expect(dispatch).not.toHaveBeenCalled();
  fireEvent(restartButton, 'press');
  expect(dispatch).toHaveBeenCalled();
});

it('should handle onCardPress', () => {
  const dispatch = jest.fn();
  const useDispatch = jest.spyOn(redux, 'useDispatch');
  useDispatch.mockReturnValue(dispatch);
  const useSelector = jest.spyOn(redux, 'useSelector');
  useSelector.mockReturnValue({
    ...initialState,
    numbers: [1, 1, 2, 2, 3, 4, 5, 5, 6, 6],
  });
  const rendered = render(<Game />);
  const oneCards = rendered.getAllByText('1');

  const firstOneCard = oneCards[0];
  expect(dispatch).not.toHaveBeenCalled();
  fireEvent(firstOneCard, 'press');
  expect(dispatch).toHaveBeenCalled();
});

it('should handle onFlipOpened', () => {
  const dispatch = jest.fn();
  const useDispatch = jest.spyOn(redux, 'useDispatch');
  useDispatch.mockReturnValue(dispatch);
  const useSelector = jest.spyOn(redux, 'useSelector');
  const mockedState = {
    ...initialState,
    numbers: [1, 2, 1, 2, 3, 4, 5, 5, 6, 6],
    firstSelected: { title: 1, id: 0 },
    secondSelected: { title: 2, id: 1 },
  };
  useSelector.mockReturnValue(mockedState);
  expect(dispatch).not.toHaveBeenCalled();
  render(<Game />);
  setTimeout(() => {
    expect(dispatch).toHaveBeenCalledTimes(3);
  }, 1000);
});
