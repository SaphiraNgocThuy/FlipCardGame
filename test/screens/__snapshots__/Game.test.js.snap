// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingTop": 44,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#2ac5db",
          "borderBottomRightRadius": 30,
          "borderTopRightRadius": 30,
          "justifyContent": "center",
          "opacity": 1,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 24,
          "paddingTop": 16,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#ffffff",
              "fontSize": 20,
              "fontWeight": "bold",
            },
          ]
        }
      >
        Restart
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "#424c58",
            "borderRadius": 30,
            "borderStyle": "solid",
            "borderWidth": 2,
            "justifyContent": "center",
            "marginRight": 16,
            "paddingBottom": 16,
            "paddingLeft": 24,
            "paddingRight": 24,
            "paddingTop": 16,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#424c58",
              "fontSize": 20,
              "fontWeight": "bold",
            },
          ]
        }
      >
        STEPS: 
        0
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignContent": "stretch",
          "flexDirection": "row",
          "flexGrow": 1,
          "flexWrap": "wrap",
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "flexBasis": "25%",
            "flexGrow": 1,
            "marginBottom": 8,
            "marginLeft": 8,
            "marginRight": 8,
            "marginTop": 8,
          },
        ]
      }
      testID="touchable"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#33c4b3",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 30,
              },
            ]
          }
        >
          1
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#424c58",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "0deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={1}
          style={
            Array [
              Object {
                "height": "100%",
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "flexBasis": "25%",
            "flexGrow": 1,
            "marginBottom": 8,
            "marginLeft": 8,
            "marginRight": 8,
            "marginTop": 8,
          },
        ]
      }
      testID="touchable"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#33c4b3",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 30,
              },
            ]
          }
        >
          1
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#424c58",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "0deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={1}
          style={
            Array [
              Object {
                "height": "100%",
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "flexBasis": "25%",
            "flexGrow": 1,
            "marginBottom": 8,
            "marginLeft": 8,
            "marginRight": 8,
            "marginTop": 8,
          },
        ]
      }
      testID="touchable"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#33c4b3",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 30,
              },
            ]
          }
        >
          2
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#424c58",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "0deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={1}
          style={
            Array [
              Object {
                "height": "100%",
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "flexBasis": "25%",
            "flexGrow": 1,
            "marginBottom": 8,
            "marginLeft": 8,
            "marginRight": 8,
            "marginTop": 8,
          },
        ]
      }
      testID="touchable"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#33c4b3",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 30,
              },
            ]
          }
        >
          2
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#424c58",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "0deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={1}
          style={
            Array [
              Object {
                "height": "100%",
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "flexBasis": "25%",
            "flexGrow": 1,
            "marginBottom": 8,
            "marginLeft": 8,
            "marginRight": 8,
            "marginTop": 8,
          },
        ]
      }
      testID="touchable"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#33c4b3",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 30,
              },
            ]
          }
        >
          3
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#424c58",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "0deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={1}
          style={
            Array [
              Object {
                "height": "100%",
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "flexBasis": "25%",
            "flexGrow": 1,
            "marginBottom": 8,
            "marginLeft": 8,
            "marginRight": 8,
            "marginTop": 8,
          },
        ]
      }
      testID="touchable"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#33c4b3",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 30,
              },
            ]
          }
        >
          4
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#424c58",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "0deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={1}
          style={
            Array [
              Object {
                "height": "100%",
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "flexBasis": "25%",
            "flexGrow": 1,
            "marginBottom": 8,
            "marginLeft": 8,
            "marginRight": 8,
            "marginTop": 8,
          },
        ]
      }
      testID="touchable"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#33c4b3",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 30,
              },
            ]
          }
        >
          5
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#424c58",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "0deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={1}
          style={
            Array [
              Object {
                "height": "100%",
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "flexBasis": "25%",
            "flexGrow": 1,
            "marginBottom": 8,
            "marginLeft": 8,
            "marginRight": 8,
            "marginTop": 8,
          },
        ]
      }
      testID="touchable"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#33c4b3",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 30,
              },
            ]
          }
        >
          5
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#424c58",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "0deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={1}
          style={
            Array [
              Object {
                "height": "100%",
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "flexBasis": "25%",
            "flexGrow": 1,
            "marginBottom": 8,
            "marginLeft": 8,
            "marginRight": 8,
            "marginTop": 8,
          },
        ]
      }
      testID="touchable"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#33c4b3",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 30,
              },
            ]
          }
        >
          6
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#424c58",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "0deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={1}
          style={
            Array [
              Object {
                "height": "100%",
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "flexBasis": "25%",
            "flexGrow": 1,
            "marginBottom": 8,
            "marginLeft": 8,
            "marginRight": 8,
            "marginTop": 8,
          },
        ]
      }
      testID="touchable"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#33c4b3",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 30,
              },
            ]
          }
        >
          6
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#424c58",
            "borderRadius": 30,
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "transform": Array [
              Object {
                "perspective": 500,
              },
              Object {
                "rotateY": "0deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={1}
          style={
            Array [
              Object {
                "height": "100%",
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    style={
      Array [
        Object {},
      ]
    }
    transparent={true}
    visible={false}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 0, 0, 0.2)",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "shadowColor": "rgba(0, 0, 0, 0.5)",
            "shadowOffset": Object {
              "height": 5,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 10,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#ffffff",
              "borderRadius": 30,
              "elevation": 10,
              "paddingBottom": 24,
              "paddingLeft": 24,
              "paddingRight": 24,
              "paddingTop": 24,
            },
          ]
        }
      >
        <Image
          resizeMode="cover"
          source={1}
          style={
            Array [
              Object {
                "height": 100,
                "width": 100,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "#424c58",
                "fontSize": 20,
                "paddingTop": 16,
              },
            ]
          }
        >
          Congratulation!!
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#424c58",
                "fontSize": 16,
                "paddingBottom": 16,
                "paddingTop": 16,
              },
            ]
          }
        >
          You win this game by 
          0
           steps
        </Text>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#2ac5db",
              "borderRadius": 30,
              "justifyContent": "center",
              "opacity": 1,
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 16,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#ffffff",
                  "fontSize": 20,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            One more game?
          </Text>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
